### PR TITLE
feat: handle non-square source images in aiig

### PR DESCRIPTION
**Pull request by jsdalton**
## Purpose

Right now, AIIG can only handle square images. But customers frequently (nearly always!) are working with non square images.

As a first step, we need to at least support square images as inputs.

## Approach

* When receiving non square images as defaults, add black bars around them to make them a square. (Later, we'll remove those black bars after we get the images back from DALL-E)

### Other details

* To make validation a bit easier, there are a few tests that now write to /tmp folder with artifacts of the images after processing. These are saved in the test (visible in artifact tab). Here's an example: https://app.circleci.com/pipelines/github/contentful/apps/23862/workflows/a1c6a058-2f0f-49bf-a894-a85a6db08ad0/jobs/70156/artifacts

#### Square

Starting image:
peaceful-cat

Starting mask:
peaceful-cat-mask

Result image:
square imageSharp

Result mask:
square maskSharp

#### Portrait

Starting image:
peaceful-cat-portrait

Starting mask:
peaceful-cat-portrait-mask

Result image:
portrait imageSharp

Result mask:
portrait maskSharp

API result:

```js
{
  "ok": true,
  "data": {
    "type": "ImageEditResult",
    "images": [
      {
        "url": "https://oaidalleapiprodscus.blob.core.windows.net/private/org-sopGxrrUI6Ap8AFV9NEMBFul/user-Cfq7XEwdtXdAZwO44IP71m5r/img-Y8kZfT6CNd8oF1PNHEDLu233.png?st=2023-09-12T12%3A56%3A40Z&se=2023-09-12T14%3A56%3A40Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2023-09-11T23%3A48%3A02Z&ske=2023-09-12T23%3A48%3A02Z&sks=b&skv=2021-08-06&sig=Qzv0PDXe83YX%2B4tXpsPwYHMYT0Yl%2BSG%2B3NI3mk8XFNg%3D",
        "imageType": "png"
      },
      {
        "url": "https://oaidalleapiprodscus.blob.core.windows.net/private/org-sopGxrrUI6Ap8AFV9NEMBFul/user-Cfq7XEwdtXdAZwO44IP71m5r/img-5rNzwzFn5qJAT46Ji8Fu2TLz.png?st=2023-09-12T12%3A56%3A40Z&se=2023-09-12T14%3A56%3A40Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2023-09-11T23%3A48%3A02Z&ske=2023-09-12T23%3A48%3A02Z&sks=b&skv=2021-08-06&sig=xymJLRy6dz3gURYjTQUsel3O08ZIg%2Bkh4qlCAWq6ito%3D",
        "imageType": "png"
      },
      {
        "url": "https://oaidalleapiprodscus.blob.core.windows.net/private/org-sopGxrrUI6Ap8AFV9NEMBFul/user-Cfq7XEwdtXdAZwO44IP71m5r/img-ecTatYePhhFIgDrlGgRSzFLT.png?st=2023-09-12T12%3A56%3A39Z&se=2023-09-12T14%3A56%3A39Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2023-09-11T23%3A48%3A02Z&ske=2023-09-12T23%3A48%3A02Z&sks=b&skv=2021-08-06&sig=goX7PXlNpXKvghgSXVprWR1r1cPVj98CbxWj0LEVSc0%3D",
        "imageType": "png"
      },
      {
        "url": "https://oaidalleapiprodscus.blob.core.windows.net/private/org-sopGxrrUI6Ap8AFV9NEMBFul/user-Cfq7XEwdtXdAZwO44IP71m5r/img-suOP1X7YF6A2scjo0jKnvQTn.png?st=2023-09-12T12%3A56%3A40Z&se=2023-09-12T14%3A56%3A40Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2023-09-11T23%3A48%3A02Z&ske=2023-09-12T23%3A48%3A02Z&sks=b&skv=2021-08-06&sig=O/E83J4ZRzG2M6O4ybcugZL5HRgogy47mSl8lpzzrZc%3D",
        "imageType": "png"
      }
    ]
  }
}
```

Example pic:
image

#### Landscape

Starting image:
peaceful-cat-landscape

Starting mask:
peaceful-cat-landscape-mask

Result image:
landscape imageSharp

Result mask:
landscape maskSharp

```js
{
  "ok": true,
  "data": {
    "type": "ImageEditResult",
    "images": [
      {
        "url": "https://oaidalleapiprodscus.blob.core.windows.net/private/org-sopGxrrUI6Ap8AFV9NEMBFul/user-Cfq7XEwdtXdAZwO44IP71m5r/img-hD5Z5LPk9LpYmz31HYAPMNzJ.png?st=2023-09-12T13%3A02%3A50Z&se=2023-09-12T15%3A02%3A50Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2023-09-11T23%3A48%3A10Z&ske=2023-09-12T23%3A48%3A10Z&sks=b&skv=2021-08-06&sig=visU3VL8jDv%2BHcWwN2WwVIrtE4rc76yBMLHRhEAKWxQ%3D",
        "imageType": "png"
      },
      {
        "url": "https://oaidalleapiprodscus.blob.core.windows.net/private/org-sopGxrrUI6Ap8AFV9NEMBFul/user-Cfq7XEwdtXdAZwO44IP71m5r/img-2m7kPmUAQHz3auZY5o4bwfSo.png?st=2023-09-12T13%3A02%3A50Z&se=2023-09-12T15%3A02%3A50Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2023-09-11T23%3A48%3A10Z&ske=2023-09-12T23%3A48%3A10Z&sks=b&skv=2021-08-06&sig=Nmhe2XXIhx9s5A13r9gMEUm6/nFvkRTauzRqzKyT3Zc%3D",
        "imageType": "png"
      },
      {
        "url": "https://oaidalleapiprodscus.blob.core.windows.net/private/org-sopGxrrUI6Ap8AFV9NEMBFul/user-Cfq7XEwdtXdAZwO44IP71m5r/img-JF4Mx4lvOJOCPRktOEmbMzGw.png?st=2023-09-12T13%3A02%3A51Z&se=2023-09-12T15%3A02%3A51Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2023-09-11T23%3A48%3A10Z&ske=2023-09-12T23%3A48%3A10Z&sks=b&skv=2021-08-06&sig=JnL5QJJswT7sir0KhDzQzjmSuurV%2BZqReOjY4ZBgHkw%3D",
        "imageType": "png"
      },
      {
        "url": "https://oaidalleapiprodscus.blob.core.windows.net/private/org-sopGxrrUI6Ap8AFV9NEMBFul/user-Cfq7XEwdtXdAZwO44IP71m5r/img-AyIEVUaEzVX6zqzmqoTJRtvr.png?st=2023-09-12T13%3A02%3A50Z&se=2023-09-12T15%3A02%3A50Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2023-09-11T23%3A48%3A10Z&ske=2023-09-12T23%3A48%3A10Z&sks=b&skv=2021-08-06&sig=S6q/JK72AkI9LDODFXAqDD3aiM002oESmJZRZsVLKN8%3D",
        "imageType": "png"
      }
    ]
  }
}
```

Example pic:

image

### CODE REVIEW

File: apps/ai-image-generator/backend/src/utils.ts [hunk 1]
```diff
@@ -19,17 +24,31 @@ export const toSharp = (imageStream: NodeJS.ReadableStream): sharp.Sharp => {
   return imageStream.pipe(sharpStream);
 };
 
+const difference = (a: number, b: number): number => {
+  return Math.abs(a - b);
+};
+
 // note: we don't care about alpha channel here, only the RGB color
```
Comment (lilbitner) [inline apps/ai-image-generator/backend/src/utils.ts:31]: nit: I suppose we can remove this note now that we added the conditional to compare alpha

**Review by lilbitner — APPROVED** [review]
Looks great to me! 

1. Terrifying creatures 💀 
2. It looks like we followed some similar logic to handling landscape/portrait images. I like how you utilized rounding with JS math, so will perhaps make some adjustments on the frontend to make it consistent with what you have here!

File: apps/ai-image-generator/backend/src/transform-images.ts [hunk 2]
```diff
@@ -128,7 +180,32 @@ export class ImageTransformer {
...
+  private toDimensions(width: number | undefined, height: number | undefined): Dimensions {
+    if (typeof width === 'undefined' || typeof height === 'undefined') {
+      throw new Error('no width or height');
+    }
+    const ratio = width / height;
+    let layout: Dimensions['layout'];
+
+    if (ratio > 1) {
+      layout = 'landscape';
+    } else if (ratio < 1) {
```
Comment (ryunsong-contentful) [inline apps/ai-image-generator/backend/src/transform-images.ts:195]: Not sure if there's ever a reality where this happens, but technically negative numbers would pass here even though it shouldn't. Maybe part of the `Math itself is broken` error should be to check for negative widths and heights?

File: apps/ai-image-generator/backend/src/transform-images.ts [hunk 3]
```diff
@@ -91,32 +73,102 @@ export class ImageTransformer {
...
-        width,
-        height,
-        fit: 'fill', // stretches to fit
+    sharpImage.resize({
+      width,
+      height,
+    });
+
+    // we only need to pad the image for portraits and landscapes -- squares are fine as is
+    if (layout === 'portrait') {
```
Comment (ryunsong-contentful) [inline apps/ai-image-generator/backend/src/transform-images.ts:99]: Food for thought: This is an example where there's repeated code but I actually think separating it out like this and having repeated code makes it easier to read even though the extraWidth calculations are all the same.

File: apps/ai-image-generator/backend/src/transform-images.ts [hunk 4]
```diff
@@ -91,32 +73,102 @@ export class ImageTransformer {
...
-    const transparentImage = await this.replaceColorWithTransparent(eraseColor, sharpImage);
-    return transparentImage.toBuffer();
+  // force the provided dimensions within the maximum side constraints
+  private constrainDimensions(dimensions: Dimensions): Dimensions {
+    const { width: startingWidth, height: startingHeight, layout, ratio } = dimensions;
+
+    let width: number;
+    let height: number;
+
+    switch (layout) {
```
Comment (ryunsong-contentful) [inline apps/ai-image-generator/backend/src/transform-images.ts:138]: I think all this switch casing is unnecessary. You can just do this one calculation for all of them regardless of the dimension layout type (Now whether it's worth keeping that type in the Dimension interface is a different story).

This computation applies for all portrait, landscapes, and sqaures. So really you can just do the below for all of them.

```
        height = Math.min(startingHeight, MAX_SIDE);
        const computedWidth = Math.round(height * ratio);
        width = Math.min(computedWidth, MAX_SIDE);
```

File: apps/ai-image-generator/backend/src/utils.ts [hunk 5]
```diff
@@ -19,17 +24,31 @@ export const toSharp = (imageStream: NodeJS.ReadableStream): sharp.Sharp => {
   return imageStream.pipe(sharpStream);
 };
 
+const difference = (a: number, b: number): number => {
```
Comment (ryunsong-contentful) [inline apps/ai-image-generator/backend/src/utils.ts:27]: Personal nit: I think this abstraction is unnecessary.

File: apps/ai-image-generator/backend/src/transform-images.ts (hunk 3 above)
Comment (jsdalton) [reply apps/ai-image-generator/backend/src/transform-images.ts:99]: Yeah I agree there. I often follow a "rule of 3" in situations like this, where I just copy and paste until I find myself doing it 3 times, in which case... time for an abstraction.

File: apps/ai-image-generator/backend/src/transform-images.ts (hunk 4 above)
Comment (jsdalton) [reply apps/ai-image-generator/backend/src/transform-images.ts:138]: Nice, this is a good insight! Let me test it out and make that change assuming it works.

Comment (jsdalton) [reply apps/ai-image-generator/backend/src/transform-images.ts:138]: Tried this and it didn't work. Realized after the fact that the order of operations matter, e.g. if you try to constraint the height of landscape first, you end up with a width well bigger than the height, which just results in a square image.

File: apps/ai-image-generator/backend/src/utils.ts (hunk 5 above)
Comment (jsdalton) [reply apps/ai-image-generator/backend/src/utils.ts:27]: Honestly it was doing my head in a bit thinking about which two values i was comparing and this helped me reason about the parent function a lot better.

**Pull request merged**